### ISSUE TITLE
Gradio App Code (Version 1)

### ISSUE THREAD

**Issue by yangyang666uwu** [yangyang666uwu/rutorrent#614]
import gradio as gr
import torch
import torchaudio
import librosa
import numpy as np
import soundfile as sf
from torch import nn
from torchaudio.transforms import MelSpectrogram

# --- Model class ---
class AudioAutoencoder(nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Conv2d(1, 16, 3, stride=2, padding=1), nn.ReLU(),
            nn.Conv2d(16, 32, 3, stride=2, padding=1), nn.ReLU()
        )
        self.decoder = nn.Sequential(
            nn.ConvTranspose2d(32, 16, 4, stride=2, padding=1), nn.ReLU(),
            nn.ConvTranspose2d(16, 1, 4, stride=2, padding=1), nn.Tanh()
        )

    def forward(self, x):
        return self.decoder(self.encoder(x))

# --- Load model ---
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = AudioAutoencoder().to(device)
model.load_state_dict(torch.load("checkpoint_epoch10.pt", map_location=device))
model.eval()

mel_transform = MelSpectrogram(sample_rate=22050, n_fft=1024, hop_length=256, n_mels=128)

# --- Enhance function ---
def enhance(input_audio):
    y, _ = librosa.load(input_audio, sr=22050)
    y = y[:22050 * 5]
    tensor_y = torch.tensor(y).unsqueeze(0)

    mel = mel_transform(tensor_y).log2().clamp(min=-10)
    mel = (mel - mel.mean()) / mel.std()
    mel = mel.unsqueeze(0).to(device)

    with torch.no_grad():
        enhanced = model(mel).squeeze().cpu().numpy()

    mel_enhanced = librosa.db_to_power(enhanced * mel.std().item() + mel.mean().item())
    out_audio = librosa.feature.inverse.mel_to_audio(mel_enhanced, sr=22050, n_fft=1024, hop_length=256)

    output_path = "enhanced_output.wav"
    sf.write(output_path, out_audio, 22050)
    return output_path

# --- Gradio UI ---
demo = gr.Interface(
    fn=enhance,
    inputs=gr.Audio(type="filepath", label="Upload Low-Quality Audio"),
    outputs=gr.Audio(type="filepath", label="Hallucinated High-Fidelity Audio"),
    title="Compassionate Audio Enhancer",
    description="Hallucinate clarity and healing into your sound. Upload a degraded file and receive a richer, more vibrant experience.",
)

demo.launch()
